### PR TITLE
[E2E]changes in e2e test for diagnostic plugin

### LINKED ISSUE #3994: [E2E] Changes in Diagnostic e2e test
## Changes Request
Current e2e test for [Diagnostic Plugin ](https://github.com/vmware-tanzu/community-edition/tree/main/cli/cmd/plugin/diagnostics/e2e-test) can have below changes.
- Currently test is running on a Minikube cluster we can run it on a Unmanaged Cluster.
- We can also transform it to a Go e2e-test using testing library which will make it platform independent(can support windows too) in future once other dependent script support every platform.

### PR DESCRIPTION
**Pull request by aman556**
## What this PR does / why we need it

This PR is doing the following changes in diagnostic e2e test.
- In place of kind cluster we have used unmanaged cluster earlier the only reason to use kind was that it is fast as we have an option to use unmanaged cluster so replaced kind with it.
- Test in written in GO to make it platform independent in future and support windows environment too for e2e tests.
- TCE installation is done by doing `make release` and to make it fast running the test on self-hosted runner.
- Other then above changes test is simulated to the previous one we can discuss in comments to improve the unchanged things.
## Which issue(s) this PR fixes

Fixes: #3994 

## Describe testing done for PR

Run `make diagnostic-e2e-test`

### CODE REVIEW

**Comment by dvonthenen** [comment]
Seems like there is an error:
https://github.com/vmware-tanzu/community-edition/runs/6085612835?check_suite_focus=true

**Comment by aman556** [comment]
Hello @vladimirvivien can you please review it once. It will be very helpful to make this test efficient.
Thanks!

**Review by vladimirvivien — COMMENTED** [review]
This looks good to me. However, will defer to, say, @dvonthenen or @jpmcb, who are more familiar with the build workflow, for a second look.

File: cli/cmd/plugin/diagnostics/Makefile
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/Makefile:23]: Do we need this target? Should probably consolidate down to just the one expected target of `e2e-test`.

File: cli/cmd/plugin/diagnostics/Makefile
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/Makefile:28]: This should call up to the root Makefile to call back down into the plugin makefile. Just have this perform the necessary calls to run the e2e tests and have the root makefile just call `make -C cli/cmd/plugin/diagnostics e2e-test`.

File: Makefile
Comment (stmcginnis) [inline Makefile:513]: See note below, but also remove passing through `$ROOT_DIR` here and just have the plugin's Makefile look it up if needed.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:11]: `ioutil` should not be used. Will have to check why the linter didn't complain about this.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:66]: May be better to replace all this with (pseudocode to just give an idea):

```
cd ../../unmanaged-cluster
go run . create clusterName
```

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:114]: Could just change to `go run . [args]`.

**Review by stmcginnis — COMMENTED** [review]
Great changes! There are a few issues, and a couple comments for possible improvements.

File: .github/workflows/e2e-diagnostics-plugin.yaml
Comment (dvonthenen) [inline .github/workflows/e2e-diagnostics-plugin.yaml:20]: For my education, did we still need these to prevent running in forks?

File: cli/cmd/plugin/diagnostics/Makefile
Comment (dvonthenen) [reply cli/cmd/plugin/diagnostics/Makefile:23]: +1

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (dvonthenen) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:11]: Yea, that's deprecated... I wonder why it didn't say anything 🤔

Comment (aman556) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:11]: Thanks @stmcginnis and @dvonthenen for reviewing. I will replace it.

File: .github/workflows/e2e-diagnostics-plugin.yaml
Comment (aman556) [reply .github/workflows/e2e-diagnostics-plugin.yaml:20]: Yeah I changed in the main workflow didn’t noticed in this one will remove it.

Comment (aman556) [reply .github/workflows/e2e-diagnostics-plugin.yaml:20]: Hey that is removed from the workflow not added if I understood your question correctly.

**Comment by aman556** [comment]
Thanks @stmcginnis and @dvonthenen for reviewing earlier. Can you folks PTAL. I have did the respective changes and resolved the conversation for now please let me know if I misunderstood any review comment.
Thanks!

File: cli/cmd/plugin/diagnostics/e2e-test/README.md
Comment (vladimirvivien) [inline cli/cmd/plugin/diagnostics/e2e-test/README.md:12]: Change `we use a Unmanaged cluster` to `we use an Unmanaged cluster`

File: .github/workflows/check-pr-diagnostics-plugin.yaml
Comment (stmcginnis) [inline .github/workflows/check-pr-diagnostics-plugin.yaml:84]: This should be removed. This is already done above in the build-release section.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:52]: nit: Should be a space after `//`.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:58]: During the "Check - Diagnostics Plugin" workflow, this causes a full build to be run twice. We should either remove the full build as a separate step from there, or add a check here first to see if `tanzu` is already present.

But that said, it looks like now we are only calling `go run` to test the code, so I don't see why we need a full install. So can we just remove this?

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (stmcginnis) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:1]: Should add tag here so the end to end tests are not run as part of running unit tests:

https://github.com/vmware-tanzu/community-edition/blob/main/cli/cmd/plugin/unmanaged-cluster/test/e2e/e2e_suite_test.go#L1

**Review by stmcginnis — COMMENTED** [review]
Thanks, this is so close. It's looking really good, just found a few more issues that I think would be good to address before we merge.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (dvonthenen) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:58]: +1

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (jpmcb) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:49]: Personal preference: I'm not a fan of these lines printing out. It goes _abit_ against the paradigm that `go test` has. If the tests has a failure, it should return that error and the `testing` library will handle it. If it succeds, there shouldn't be anything printed to the screen: just a simple `ok` which is also handled by the Go `testing` standard library.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (jpmcb) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:58]: +1

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (jpmcb) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:70]: Calling `go run` here could be replaced by importing our library:

```go
import (
...
    github.com/vmware-tanzu/community-edition/cli/cmd/plugin/unmanaged-cluster
)
```

And than using the well defined `Manager.Deploy` API in the `tanzu` package to get a cluster spun up.

It essentially does the same thing as calling `./unmanaged cluster create` from the command line but without having to worry about system paths and using os.Exec

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (jpmcb) [inline cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:145]: Same here. We could easily import the `unmanaged-cluster` tanzu package and call `Manager.Delete` with the cluster name.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (jpmcb) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:70]: But, in the short term, I'm ok bypassing this change for now and opening a new issue to move towards that mode so large chunks of this don't have to be re-written.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (aman556) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:58]: There is check for `tanzu` cli in diagnostic code [here](https://github.com/vmware-tanzu/community-edition/blob/aa4549af85e8073ddec564ed9a9025b9831de696/cli/cmd/plugin/diagnostics/pkg/collect_cmd.go#:~:text=gexe.Prog().Avail(%22-,tanzu,-%22)%20%3D%3D%20%22%22%20%7B) due to which we need tanzu.

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (aman556) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:70]: I will add this comment link in the code as comment for future improvement.
Thanks!

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (aman556) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go:49]: Hey @jpmcb I also gave a thought, agree what you mentioned above. The only motive was to make the logs user friendly but as verbosity is not that much and we can have other options too so I am  removing them for now.
Thanks!

File: .github/workflows/check-pr-diagnostics-plugin.yaml
Comment (aman556) [reply .github/workflows/check-pr-diagnostics-plugin.yaml]: strange but I think for different jobs we have to do different setup i removed that thing and workflow failed https://github.com/vmware-tanzu/community-edition/runs/6205983295?check_suite_focus=true#step:2:9

Comment (aman556) [reply .github/workflows/check-pr-diagnostics-plugin.yaml]: upto my knowledge jobs run on separate machine 
[found this in workflow doc](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#:~:text=Each%20job%20will%20run%20inside%20its%20own%20virtual%20machine%20runner%2C%20or%20inside%20a%20container%2C%20and%20has%20one%20or%20more%20steps%20that%20either%20run%20a%20script%20that%20you%20define%20or%20run%20an%20action%2C%20which%20is%20a%20reusable%20extension%20that%20can%20simplify%20your%20workflow.)

**Comment by aman556** [comment]
I have resolved the conversation which were direct had left some of them open which might need discussion. For the scenario the changes for the open conversation are satisfied I am squashing the commits. Only one check is failing which is a know issue.
Latest logs - https://github.com/vmware-tanzu/community-edition/runs/6206320431?check_suite_focus=true#:~:text=1-,Run,-make%20diagnostic%2De2e
Thanks!   
CC @stmcginnis @jpmcb @dvonthenen @vladimirvivien

File: .github/workflows/check-pr-diagnostics-plugin.yaml
Comment (stmcginnis) [reply .github/workflows/check-pr-diagnostics-plugin.yaml:84]: We should make this one job with multiple steps then. I don't see a strong reason to have two separate jobs if it adds redundancy. But that's something we can follow up on later. Thanks!

**Review by stmcginnis — APPROVED** [review]
Awesome, things are looking pretty good. Thanks for addressing all of the comments.

Will wait to merge so others get a chance to review.

Nice work!

**Review by davidvonthenen — APPROVED** [review]

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (dvonthenen) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go]: I think we should remove that `tanzu` binary check that you referenced.

We should actually do the second option:
`remove the full build as a separate step from there` as it is the only way for us to trigger and consume from the nightly build

**Review by davidvonthenen — DISMISSED** [review]
I left this comment...

https://github.com/vmware-tanzu/community-edition/pull/4106#discussion_r861058786

File: cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go
Comment (aman556) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go]: Hey @dvonthenen just to clarify you mean removing the tanzu build code from the e2e-test and adding it in workflow? like [this](https://github.com/vmware-tanzu/community-edition/blob/main/.github/workflows/check-pr-diagnostics-plugin.yaml#:~:text=instance%2Did%3A%3A%24%7BINSTANCE_NAME%7D-,build%2Drelease%3A,-name%3A%20Build%20Release)?

Comment (dvonthenen) [reply cli/cmd/plugin/diagnostics/e2e-test/e2e_test.go]: Yea... I can see how that is confusing. We should remove the build from the workflow and the download-or-build-sh script should do the build.

**Review by davidvonthenen — APPROVED** [review]

**Pull request merged**
